enable click on panel icons to toggle lock keys

samlane-ma opened the pull request:
## Add ability to toggle lock keys via applet
- Allows clicking on lock key applet icons to toggle the state, using xdotool if present
- This will not connect the signals if xdotool is not available to avoid  forcing additional dependencies on Solus

### Submitter Checklist

- [x] Squashed commits with `git rebase -i` (if needed)
- [x] Built budgie-desktop and verified that the patch worked (if needed)

samlane-ma commented:
Addresses #1791

JoshStrobl commented on src/applets/lock-keys/LockKeysApplet.vala:
```diff
@@ -72,6 +88,25 @@ public class LockKeysApplet : Budgie.Applet {
+	/* If xdotool is installed, handle the clicks on the panel icons */
+	protected bool on_panel_widget_clicked(Gdk.EventButton eventbutton, string button) {
+		if (eventbutton.button == 1)
+		{
```
curly brace should be after the if statement / condition, not on a new line.

JoshStrobl commented on src/applets/lock-keys/LockKeysApplet.vala:
```diff
@@ -19,7 +19,10 @@ public class LockKeysApplet : Budgie.Applet {
+	Gtk.EventBox caps_box;
+	Gtk.EventBox num_box;
 	new Gdk.Keymap map;
+	const string XDOTOOL = "/usr/bin/xdotool";
```
Should not be hard coded. See other comment re. using Environment

JoshStrobl commented on src/applets/lock-keys/LockKeysApplet.vala:
```diff
@@ -72,6 +88,25 @@ public class LockKeysApplet : Budgie.Applet {
+	protected bool on_panel_widget_clicked(Gdk.EventButton eventbutton, string button) {
+		if (eventbutton.button == 1)
+		{
+			try {
```
Should avoid calling if we don't have it in PATH. See https://valadoc.org/glib-2.0/GLib.Environment.find_program_in_path.html, if it is null then it shouldn't try spawning the process.

samlane-ma replied:
Thanks for feedback. Could I get a clarification please on this point (in case I am misunderstanding.)  The on_panel_widget_clicked callback function doesn't get connected to the button_press_event(s) if xdotool is not found, so without xdotool, this function should never be called.  I didn't add a check here because it felt redundant, but easy enough to add for good measure, if recommended.

JoshStrobl replied:
I don't think it would hurt for us to set up the event listeners anyways and just return the event propagation if we don't have xdotool installed. It wouldn't make a difference to the user and we would get the benefit of the user being able to install xdotool whenever (assuming we don't force it as a rundep) and it would just immediately start working, since `GLib.Environment` would be able to find it in PATH. I don't really expect any meaningful performance regressions resulting from checking it whenever we click it, realistically the user isn't going to be doing it all that often.

Hope that makes sense :)

samlane-ma commented:
Made changes as suggested:
- GLib.Environment is now being used to determine the path
- Listeners are being set up regardless of xdotool now
- Moved the xdotool check to the callback function, and ensure it won't try to spawn if not in path
- Fixed formatting error

Tested installing xdotool with applet already running, now it indeed does start working immediately
Tested uninstalling xdotool with applet already running just to make sure there are no surprises

Thank you, your time (and patience) is appreciated.

JoshStrobl commented on src/applets/lock-keys/LockKeysApplet.vala:
```diff
@@ -19,7 +19,10 @@ public class LockKeysApplet : Budgie.Applet {
+	Gtk.EventBox caps_box;
+	Gtk.EventBox num_box;
 	new Gdk.Keymap map;
+	string? xdotool;
```
This doesn't actually need to be a class property, it can just be a variable limited to the scope of the `on_panel_widget_clicked` function, e.g. `string? xdotool = Environment.find_program_in_path("xdotool");`

Building on that, instead of having the nesting of the try / catch, we can just do:

``` vala
if ((xdotool == null) || (eventbutton.button != 1)) {
	return Gdk.EVENT_PROPAGATE;
}
```

Then have the try / catch

samlane-ma replied:
Thanks.  For the "catch", on the off-chance of a spawn error, is it preferred to propagate or stop propagation?

samlane-ma commented:
Made the changes requested.
Again, thank you!

JoshStrobl reviewed (approved):
Patch looks good to me. Thanks for the improvements!
